store/sagas/provider.tsx

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {useCallback, useMemo, useRef} from 'react';
+import {Saga, SagaMiddleware, Task} from 'redux-saga';
+
+import {SagaRunnerContext} from './context';
+
+export type Props = {
+    children: React.ReactNode;
+    sagaMiddleware: SagaMiddleware;
+}
+
+export function SagaProvider(props: Props) {
```

Comment:
Currently, this provider gives components access to two functions: `isSagaRunning` and `startSaga`. Both of those require a unique name to avoid accidentally running two copies of the same saga when called from places like a data manager hook.

I think `isSagaRunning` is straightforward enough. It's intended to be used before calling `startSaga` since I decided to make `startSaga` fail when called twice instead of letting someone accidentally run two copies of a saga at once. It's possible that there might be some use cases for that behaviour, but I think we can deal with that if it comes up in practice.

As its name suggests, `startSaga` starts a saga. It doesn't provide any guarantees that that saga will keep running because sagas that error out will immediately end, but perhaps that could be added in the future. Either that or we could at least have it clean up `runningSagas` automatically. It also returns a function that can be used to stop the saga in case a component wants to do that on unmount, but I don't think that will always be necessary since we may just want to lazily initialize background sagas for things like the data manager.

As an example for how this might look in use, this is a version of the `useUser` data manager hook that I've been playing with:

```ts
export function useUser(userId: string) {
    const sagaRunner = useContext(SagaRunnerContext);
    const dispatch = useDispatch();

    const user = useSelector((state: GlobalState) => getUser(state, userId));

    if (user) {
        return user;
    }

    if (!sagaRunner.isSagaRunning('useUser')) {
        sagaRunner.startSaga('useUser', userSaga);
    }

    dispatch({
        type: 'FETCH_USER_BY_ID',
        userId,
    });

    return undefined;
}
```

There might be some improvements that could be made to usability here such as adding a dedicated hook for this instead of using `useContext` or having some sort of `startIfNotRunning` function. I don't think this exactly needs a super easy-to-use interface though since I don't expect this will be used in this form in many places. I'll likely end up wrapping the data manager part in some helper stuff to hide the sagas from someone adding their own hook for it.